### PR TITLE
hillst/bionemo noodles

### PR DESCRIPTION
**Pull request by skothenhill-nv**
# NvFaidx
Adds a memmapped faidx reader with python bindings.

## Context 
PyFaidx or _any_ buffered read based index is not process safe, and therefore does not play nice with pytorch dataloaders.

Due to the order of operations, the underlying file handle is shared between processes, when `seek()` is called to perform random lookups, this can cause unexpected behavior in the forked processes.
Ref: https://github.com/mdshw5/pyfaidx/issues/211

For a good solution we need three things:
    1) Safe index creation, in multi-process or multi-node scenarios, this should be restricted to a single node where all workers block until it is complete (not implemented above)
    2) Index object instantion must be fast.
    3) Read-only use of the index object must be both thread safe and process safe with python.
    
## Memmap backing
Using a backend of Memmaps provide the following benefits (Credit: @malcolmgreaves)
- The kernel implements this mechanism by using page faults
- Each read in a mmap'd file results in a page fault: there's nothing in memory to read!
- The kernel handles this page fault by going to the disk, reading the file in the specified offset + index, then returning to the user process with what it just read, preventing penalties from context switching.

These three usecases are ideal for pytorch DataLoaders!

## Perf Benchmarks

nvfaidx query faster by 5.2x
nvfaidx instantiation faster by 16x

Metric | pyfaidx (s) | nvfaidx (s)
-- | -- | --
Query/s | 1.668e-05 | 3.239e-06
Instantiation Time | 0.01001 | 0.0006313

## TODO:
- Dockerfile needs to be reworked to play nice with cargo.

## Usage:

Import NvFaidx, pass in your fasta file path, and go fast.
```
from bionemo.noodles import NvFaidx
index = NvFaidx('sub-packages/bionemo-noodles/tests/bionemo/noodles/data/sample.fasta')
assert index['chr1'][0] == 'A'
assert index['chr1'][1:4] == 'CTG'
assert index['chr1'][1:10000] == 'CTGACTGACTG'
assert index['chr1'][0:-1] == 'ACTGACTGACT'
assert index['chr1'][:-1] == 'ACTGACTGACT'
assert index['chr1'][-1:] == 'G'
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -172,6 +172,12 @@ EOF
 
 FROM dev AS development
 
+# TODO (SKH): This doesnt seem safe!
+RUN curl https://sh.rustup.rs -sSf | sh -s -- -y && \
+    source $HOME/.cargo/env && \
+    rustup default nightly
```
Comment (pstjohn) [inline Dockerfile:178]: this would be a great place for a multi-stage build. Can we have a separate rust base image that builds the libraries and then we just copy them over into the release container

**Review by malcolmgreaves — COMMENTED** [review]
First parital review.

File: Dockerfile [hunk 2]
```diff
@@ -165,48 +170,51 @@ RUN <<EOF
   pip uninstall -y nemo_toolkit megatron_core
 EOF
 
+############## NOTE this was all 'removed' for some reason
```
Comment (jstjohn) [inline Dockerfile:173]: is this necessary?

File: Dockerfile [hunk 3]
```diff
@@ -91,13 +91,18 @@ WORKDIR /workspace/bionemo2
 COPY ./3rdparty /workspace/bionemo2/3rdparty
 COPY ./sub-packages /workspace/bionemo2/sub-packages
 
+RUN ls -R /workspace/bionemo2/sub-packages/bionemo-noodles
```
Comment (jstjohn) [inline Dockerfile:94]: What does this get us?

File: Dockerfile [hunk 4]
```diff
@@ -165,48 +170,51 @@ RUN <<EOF
...
-
 FROM dev AS development
 
 WORKDIR /workspace/bionemo2
 COPY --from=bionemo2-base /workspace/bionemo2/ .
 COPY ./internal ./internal
 # because of the `rm -rf ./3rdparty` in bionemo2-base
 COPY ./3rdparty ./3rdparty
+
+########### End note
```
Comment (jstjohn) [inline Dockerfile:187]: why the end note?

**Comment by polinabinder1** [comment]
Are the perf benchmarks for a random query? Are they averaged?

File: sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (jstjohn) [inline sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py:1]: Love these tests!

**Comment by skothenhill-nv** [comment]
> Are the perf benchmarks for a random query? Are they averaged?

1000 random queries on a randomly generated sequence. Code for benchmarking is available in the unit tests

File: Dockerfile [hunk 6]
```diff
@@ -88,16 +88,21 @@ RUN --mount=type=bind,source=./sub-packages/bionemo-geometric/requirements.txt,t
...
 # Install 3rd-party deps and bionemo submodules.
-COPY ./3rdparty /workspace/bionemo2/3rdparty
-COPY ./sub-packages /workspace/bionemo2/sub-packages
+COPY ./ /workspace/bionemo2
+#COPY ./3rdparty /workspace/bionemo2/3rdparty
+#COPY ./sub-packages /workspace/bionemo2/sub-packages
+
+RUN --mount=type=cache,target=/root/.cache,sharing=locked curl https://sh.rustup.rs -sSf | sh -s -- -y
+ENV PATH="/root/.cargo/bin:${PATH}"
+RUN rustup default nightly
```
Comment (malcolmgreaves) [inline Dockerfile:97]: Pin this to a specific rust toolchain version.

File: Dockerfile [hunk 7]
```diff
@@ -88,16 +96,23 @@ RUN --mount=type=bind,source=./sub-packages/bionemo-geometric/requirements.txt,t
 WORKDIR /workspace/bionemo2
 
 # Install 3rd-party deps and bionemo submodules.
+COPY ./LICENSE /workspace/bionemo2/LICENSE
```
Comment (trvachov) [inline Dockerfile:99]: 👏

File: Dockerfile [hunk 8]
```diff
@@ -88,16 +96,23 @@ RUN --mount=type=bind,source=./sub-packages/bionemo-geometric/requirements.txt,t
...
 COPY ./3rdparty /workspace/bionemo2/3rdparty
 COPY ./sub-packages /workspace/bionemo2/sub-packages
 
+COPY --from=rust-env /usr/local/cargo /usr/local/cargo
+COPY --from=rust-env /usr/local/rustup /usr/local/rustup
+
+ENV PATH="/usr/local/cargo/bin:/usr/local/rustup/bin:${PATH}"
+ENV RUSTUP_HOME="/usr/local/rustup"
```
[the remark this answers is not in the document]
Comment (skothenhill-nv) [reply Dockerfile:107]: definitely need it here or those pip installs will fail. do you have another suggestion? we could remove them from the release image at the end of the Dockerfile perhaps?

File: sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        if isinstance(fasta_path, Path):
+            fasta_path = str(fasta_path)
+        if isinstance(faidx_path, Path):
+            faidx_path = str(faidx_path)
+        if ignore_existing_fai:
+            self.reader = PyIndexedMmapFastaReader(fasta_path, ignore_existing_fai=ignore_existing_fai)
+        elif faidx_path is not None:
+            self.reader = PyIndexedMmapFastaReader.from_fasta_and_faidx(fasta_path, faidx_path)
+        else:
+            self.reader = PyIndexedMmapFastaReader(fasta_path)
```
Comment (edawson) [inline sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:128]: I think you can reduce this to:

```
if not ignore_existing_fai and faidx_path is not None:
    self. reader = PyIndexedMmapFastaReader.from_fasta_and_faidx(fasta_path, faidx_path)
else:
    self.reader = PyIndexedMmapFastaReader(fasta_path, ignore_existing_fai=ignore_existing_fai)
```

It's up to opinion which one is clearer, as they're both special cases that are largely determined by where errors would get thrown, but something to consider.

File: sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    Ref: https://github.com/mdshw5/pyfaidx/issues/211
+
+    For a good solution we need three things:
+        1) Safe index creation, in multi-process or multi-node scenarios, this should be restricted to a single node
+            where all workers block until it is complete (not implemented above)
+        2) Index object instantion must be fast.
+        3) Read-only use of the index object must be both thread safe and process safe with python.
+    """
+
+    def __init__(self, fasta_path: str | Path, faidx_path: Optional[str | Path] = None, ignore_existing_fai=True):
```
Comment (edawson) [inline sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:110]: Nit: add type hint to ignore_existing_fai

```python
def __init__(self, fasta_path: str | Path, faidx_path: Optional[str | Path] = None, ignore_existing_fai: bool =True)
```

File: Dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (skothenhill-nv) [reply Dockerfile:107]: its needed for the pip install here unless we figure out how to compile/link manually with maturin/pyo3

File: sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py (hunk 10 above)
Comment (edawson) [inline sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:110]: Nit: Should this be `str | path`, or some variant of `Union[str, Path, None]` for both fasta_path and faidx_path?

**Review by farhadrgh — APPROVED** [review]

File: sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        """Construct a dict-like object representing a memmapped, indexed FASTA file.
+
+        Args:
+            fasta_path (str): Path to the FASTA file.
+            faidx_path (str): Path to the FAI index file. If None, one will be created.
+            ignore_existing_fai (bool): If True, ignore any existing FAI file and create an in-memory index. Note that
+                this will also ignore `faidx_path`.
+        """
+        if isinstance(fasta_path, Path):
+            fasta_path = str(fasta_path)
```
Comment (edawson) [inline sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:120]: Should we explicitly check that this value is either string or path and throw an error if it is not here?

**Review by pstjohn — APPROVED** [review]

File: sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+    filename = create_test_fasta(num_seqs=2, seq_length=200)
+    faidx_filename = PyIndexedMmapFastaReader.create_faidx(filename)
+    index = PyIndexedMmapFastaReader.from_fasta_and_faidx(filename, faidx_filename)
+    index2 = PyIndexedMmapFastaReader(filename, ignore_existing_fai=True)
+    # Test against constructor for equivalence.
+    assert index.read_sequence_mmap("contig1:1-1") == index2.read_sequence_mmap("contig1:1-1")
+    assert index.read_sequence_mmap("contig1:1-10") == index2.read_sequence_mmap("contig1:1-10")
+    assert index.read_sequence_mmap("contig1:10-200") == index2.read_sequence_mmap("contig1:10-200")
+
+    # no we are going to rename the file, and then try again, we expect the same outcome.
```
Comment (edawson) [inline sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py:69]: Nit: "no" -> "now"

File: Dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (skothenhill-nv) [reply Dockerfile:107]: OK, did the cleanup thing at the end that doesnt really do much. Also remove the `target/` folder thats created during the first pip install as we discussed on slack.

File: sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+    assert index.read_sequence_mmap("chr4:1-17") == "CCCCCCCCCCCCACGT"
+    assert index.read_sequence_mmap("chr4:1-1000000") == "CCCCCCCCCCCCACGT"
+
+    assert index.read_sequence_mmap("chr4:1-17") == "CCCCCCCCCCCCACGT"
+
+    assert index.read_sequence_mmap("chr4:3-16") == "CCCCCCCCCCACGT"
+    assert index.read_sequence_mmap("chr4:17-17") == ""
+
+
+def test_getitem_bounds(sample_fasta):
```
Comment (edawson) [inline sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py:132]: Please add a test for invalid contig.

File: sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+        return 99999999999
+
+    def __getitem__(self, idx):
+        # Always return the same thing to keep it easy, we assume the fasta_created is doing the right thing.
+        return str(self.fasta["contig1"][150000:160000])
+
+
+@pytest.mark.skip
+@pytest.mark.xfail(reason="This is a known failure mode for pyfaidx that we are trying to prevent with nvfaidx.")
+def test_parallel_index_creation_pyfaidx():
```
Comment (edawson) [inline sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py:229]: Suggest adding iteration time tests here.

File: sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py [hunk 15]
```diff
@@ -0,0 +1,403 @@
...
+    for i, batch in enumerate(dl):
+        if i > max_i:
+            break
+        lens = [len(x) for x in batch]
+        lens_equal = [x == 10000 for x in lens]
+        assert all(lens_equal), (set(lens), sum(lens_equal))
+
+
+def test_file_errors():
+    # test missing fasta file
```
Comment (edawson) [inline sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py:271]: Please test missing FAIDX file as well.

**Review by edawson — COMMENTED** [review]
Some changes requested, but overall in good shape. Recommend refactoring PyRecord to a more descriptive name, adding some small tests, and cleaning up a little branchiness. I also noticed a lot of errors get pushed down to the rust class - is this desired?

File: sub-packages/bionemo-fw/pyproject.toml [hunk 16]
```diff
@@ -17,6 +17,7 @@ dependencies = [
     'bionemo-geneformer',
     'bionemo-geometric',
     'bionemo-llm',
+    'bionemo-noodles',
```
[the remark this answers is not in the document]
Comment (skothenhill-nv) [reply sub-packages/bionemo-fw/pyproject.toml:20]: Sort of- its required for building bionemo-noodles from source, but if the compiled shared object file is there, theoretically the wheel can still be built.

File: sub-packages/bionemo-noodles/rust/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,662 @@
...
+/// Python wrapper around the faidx Record struct.
+/// Fields:
+/// - name: name of the record, corresponds to a sequence id in the indexed fasta file.
+/// - length: length of the record, number of bases/nucleotides/characters in the record, including Ns.
+/// - offset: offset of the record's first base/nucleotide/character in bytes, from the start of the file.
+/// - line_bases: number of bases per line in the fasta file
+/// - line_width: number of bytes per line in the fasta file, including newlines, return carriages, etc.
+#[pyclass]
+#[derive(Clone)]
+struct PyRecord {
```
[the remark this answers is not in the document]
Comment (skothenhill-nv) [reply sub-packages/bionemo-noodles/rust/src/lib.rs:19]: I'm copying the semantics from Noodles-rs, does that change your opinion?

File: sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+    *Context*: PyFaidx or _any_ buffered read based index is not process safe, and therefore does not play nice with pytorch dataloaders.
+    Due to the order of operations, the underlying file handle is shared between processes, when `seek()` is called to perform random lookups,
+    this can cause unexpected behavior in the forked processes.
+    Ref: https://github.com/mdshw5/pyfaidx/issues/211
+
+    For a good solution we need three things:
+        1) Safe index creation, in multi-process or multi-node scenarios, this should be restricted to a single node
+            where all workers block until it is complete (not implemented above)
+        2) Index object instantion must be fast.
+        3) Read-only use of the index object must be both thread safe and process safe with python.
```
[the remark this answers is not in the document]
Comment (skothenhill-nv) [reply sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:107]: I got the opposite feedback in a different review, I'm going to keep them for now.

File: sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py (hunk 10 above)
Comment (skothenhill-nv) [reply sub-packages/bionemo-noodles/src/bionemo/noodles/nvfaidx.py:110]: nope! if `faidx_path` is None and `ignore_existing_fai` is True, its perfectly valid, the index is built online and stored in memory in this case. (this is in the docstring too I believe). 

But fasta_path must be set.

File: sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py (hunk 14 above)
Comment (skothenhill-nv) [reply sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py:229]: discussed offline- this is meant for the DataLoader usecase.

File: sub-packages/bionemo-noodles/rust/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,662 @@
...
+    fn __str__(&self) -> String {
+        format!(
+            "PyRecord(name={}, length={}, offset={}, line_bases={}, line_width={})",
+            self.name, self.length, self.offset, self.line_bases, self.line_width
+        )
+    }
+
+    fn __repr__(&self) -> String {
+        format!(
+            "<PyRecord name={} length={} offset={} line_bases={} line_width={}>",
```
[the remark this answers is not in the document]
Comment (skothenhill-nv) [reply sub-packages/bionemo-noodles/rust/src/lib.rs:62]: doneeeee

File: sub-packages/bionemo-noodles/rust/src/lib.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (skothenhill-nv) [reply sub-packages/bionemo-noodles/rust/src/lib.rs:19]: changed it anyway, i agree.

File: sub-packages/bionemo-noodles/VERSION [hunk 20]
```diff
@@ -0,0 +1 @@
+../../VERSION
```
[the remark this answers is not in the document]
Comment (skothenhill-nv) [reply sub-packages/bionemo-noodles/VERSION:1]: I'm not sure, im going with status quo here.
